### PR TITLE
gh-92248: Deprecate `type`, `choices`, `metavar` parameters of `argparse.BooleanOptionalAction`

### PR DESCRIPTION
**Pull request by sobolevn**
`argparse.BooleanOptionalAction` was added in 3.9
https://github.com/python/cpython/issues/85039 removed just a single wrong parameter.
But, it was on time when 3.9 was recently released, and no deprecation warning was issued.
The parameter was just removed.

But, since we already have 3.9, 3.10, and 3.11, (and all alphas of 3.12), I am sure that we need a deprecation period now.

People might use this (the use-case is probably wrong), but I don't like breaking things for no clear benefit.

I'm feeling pretty safe about this deprecation (I even considered just removing these parameters straight away), because code search shows that no one is using these arguments: https://cs.github.com/?q=action%3DBooleanOptionalAction

Plus, they don't make any sense and are mostly not changing any behaviour of `BooleanOptionalAction`.

CC @barneygale since you've asked for this 😆 
CC @rhettinger and @hauntsaninja  since you've reviewed https://github.com/python/cpython/issues/85039

* Issue: gh-92248

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2023-04-22-12-30-10.gh-issue-92248.NcVTKR.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Deprecate ``type``, ``chocies``, and ``metavar`` parameters of
```
Comment (Eclips4) [inline Misc/NEWS.d/next/Library/2023-04-22-12-30-10.gh-issue-92248.NcVTKR.rst]:
```suggestion
Deprecate ``type``, ``choices``, and ``metavar`` parameters of
```

Comment (sobolevn) [reply Misc/NEWS.d/next/Library/2023-04-22-12-30-10.gh-issue-92248.NcVTKR.rst]: Thanks, tpyos are my thing :)

**Comment by rhettinger** [comment]
This would gradually nudge the code toward the way it should have been designed from the outset. Eventually, it might save someone a few minutes of time.  

That said, this is a pretty minor issue (ignoring unused arguments). Changing the behavior will be inconvenient for people who happened to have fed in an unused argument in code that is currently functioning correctly — we've breaking code that was sloppy but correct — I'm not sure they will appreciate being forced to make a minor edit that doesn't change behavior or fix any problem that they have.

**Comment by sobolevn** [comment]
@rhettinger thank you for your input. However, I don't quite agree with some of your points.

> Changing the behavior will be inconvenient for people who happened to have fed in an unused argument in code that is currently functioning correctly

Two things here:
1. `Changing the behavior will be inconvenient`: Yes, it can cause a new warning. But, this is pretty easy to deal with: just remove the argument. It won't require any compat code or migration plan. I think that this is a good thing, because it helps to indicate a problem with your code. And, of course, people with the correct code won't have any warnings at all
2. `currently functioning correctly`: I don't think that we can call this code "correct". Here's the first example:

```python
>>> from argparse import *
>>> parser = ArgumentParser()
>>> parser.add_argument('--foo', action=BooleanOptionalAction, type=int)  # we expect `int`
>>> parser.parse_args(['--foo'])  # but get `bool`
Namespace(foo=True)
```

Here's the code does not "function correctly" in a sense that `type` is ignored. While users explicitly asks to use `int`.

The second example:

```python
>>> from argparse import *
>>> parser = ArgumentParser()
>>> parser.add_argument('--foo', action=BooleanOptionalAction, choices=['on', 'off'])
>>> parser.parse_args(['--foo'])
Namespace(foo=True)
```

And again: user specifies two values that must be respsected. But, they are ignored.

We also do not document this behaviour:
Снимок экрана 2023-04-29 в 12 59 28

So, any user might expect it to work similarly to other action types. Which is not the case right now.

All in all, I see this as an improvement:
1. We show what's wrong and do not break any code
2. In the future we will have the correct signature that does the correct thing

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -730,6 +730,12 @@ Pending Removal in Python 3.14
 * The *onerror* argument of :func:`shutil.rmtree` is deprecated in 3.12,
   and will be removed in 3.14.
 
+* *type*, *chocies*, and *metavar* parameters
```
Comment (hugovk) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* The *type*, *choices*, and *metavar* parameters
```

File: Doc/whatsnew/3.12.rst [hunk 3]
```diff
@@ -730,6 +730,12 @@ Pending Removal in Python 3.14
...
+  of :class:`!argparse.BooleanOptionalAction` are deprecated
+  and will be removed in 3.14
```
Comment (hugovk) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  and will be removed in 3.14.
```

File: Lib/argparse.py [hunk 4]
```diff
@@ -901,6 +904,24 @@ def __init__(self,
...
+                    "{name!r} is deprecated as of Python 3.12 and will be "
+                    "removed in Python {remove}.",
+                    remove=(3, 14))
+
+        if type is _deprecated_default:
+            type = None
+        if choices is _deprecated_default:
+            choices = None
+        if metavar is _deprecated_default:
+            metavar = None
```
Comment (hugovk) [inline Lib/argparse.py:924]: Could we remove these three if checks and reset all three back to `None`, with a comment?

Comment (sobolevn) [reply Lib/argparse.py:924]: It feels safer not to do that. Right now my change does not change any behavior (for subclasses, etc). But, setting these 3 params to `None` without conditions might change it for some corner cases. So, I would rather not :)

**Review by hugovk — APPROVED** [review]

**Comment by hugovk** [comment]
Let's merge this before the beta if there's no remaining objections.

**Comment by actualben** [comment]
Hi from the future (post 2025 Q4 when the params were removed). For the record, there was a reason those parameters were accepted in the first place and that reason seems totally unexamined here. Also the codesearch described missed the mark quite a bit (for example, as of 3-Feb-2026):

* <https://github.com/search?type=code&q=action%3DBooleanOptionalAction> - 1.3k files
* vs <https://github.com/search?type=code&q=action%3Dargparse.BooleanOptionalAction> - **14.6k files**

And this change is documented in the wrong place -- people don't instantiate `argparse.BooleanOptionalAction` directly (which is proven in the docs for that class: <https://docs.python.org/3.14/library/argparse.html#argparse.BooleanOptionalAction>), instead they call `add_argument` with the action param set accordingly, but <https://docs.python.org/3.14/library/argparse.html#choices> shows nothing about `choices=None` (for example) suddenly being a fatal error. You might want to do something like `action=BooleanOptionalAction, choices=var_that_evaluates_to_none` when you're calling `add_argument` dynamically. There are workarounds of course, but this breakage seems unforced - when we could have just warned about questionable code without changing the API:

```python
if type is not None:
  warnings._whatever("type is implied when action is BooleanOptionalAction")
if choices is not None:
  warnings._whatever("choices are predefined with BooleanOptionalAction")
if metavar is not None:
  warnings._whatever("there is no metavar when action is BooleanOptionalAction")
```

**Pull request merged**